src/jekyll/jk_document.rs

```diff
@@ -0,0 +1,219 @@
+                v.as_array().map(
+                    |v| v.iter().map(|v| v.to_string()).collect(),
+                )
+            }))
+            .merge_permalink(custom_attributes.remove("permalink").and_then(|v| {
+                v.as_str().map(|s| s.to_owned())
+            }))
+            .merge_draft(custom_attributes.remove("published").and_then(|v| {
+                v.as_bool().map(|b| !b)
+            }))
+            .merge_layout(custom_attributes.remove("layout").and_then(|v| {
+                v.as_str().map(|s| s.to_owned())
+            }))
+            .merge_published_date(custom_attributes.remove("date").and_then(|d| {
+                d.as_str().and_then(datetime::DateTime::parse)
+            }))
+            .merge_custom(custom_attributes)
+    }
+}
+
+#[derive(Debug, Default, Clone, Eq, PartialEq)]
+#[derive(Serialize, Deserialize)]
+pub struct JkDocument {
+    pub front: Option<String>,
+    pub content: Option<String>,
+}
+
+impl JkDocument {
+    pub fn parse_string(doc: String) -> Result<JkDocument> {
+        let (front, content) = split_document(&doc)?;
+        if front != None {
+            Err(ErrorKind::MissingFrontStart.into())
+        } else {
+            match split_document(content)? {
+                (None, _) => Err(ErrorKind::MissingFrontmatter.into()),
+                (Some(front), content) => Ok(JkDocument {
+                    front: Some(front.to_owned()),
+                    content: Some(content.to_owned()),
+                }),
+            }
+        }
+    }
+
+    pub fn parse(source_file: &Path) -> Result<JkDocument> {
+        let doc: String = read_file(source_file)?;
+        JkDocument::parse_string(doc)
+    }
+
+    pub fn convert_front(front: Option<String>) -> Result<String> {
+        let front_value: JkFrontmatterBuilder = front
+            .map(|s| serde_yaml::from_str(&s))
+            .map_or(Ok(None), |r| r.map(Some))?
+            .unwrap_or_else(JkFrontmatterBuilder::new);
+
+        let front_builder: FrontmatterBuilder = front_value.into();
+        let front = front_builder.build()?;
+        let mut converted = serde_yaml::to_string(&front)?;
+        converted.drain(..4);
+        Ok(converted)
+    }
+
+    pub fn convert(source_file: &Path, dest_dir: &Path) -> Result<()> {
+        let doc = JkDocument::parse(source_file)?;
+        let front = JkDocument::convert_front(doc.front)?;
+        if !dest_dir.exists() {
+            create_dir_all(&dest_dir)?;
+        }
+        let dest_file = dest_dir.join(source_file.file_name().unwrap());
+        let mut dest = File::create(dest_file)?;
+        let converted = format!("{}\n---\n{}", &front, &doc.content.unwrap());
+        dest.write_all(converted.as_bytes())?;
+        Ok(())
+    }
+}
+
+pub fn convert_from_jk(source: &Path, dest: &Path) -> Result<()> {
+    if dest.is_file() {
+        Err(ErrorKind::CantOutputInFile.into())
+    } else if source.is_file() {
+        JkDocument::convert(source, dest)
```

Comment:
Should this check the file type and convert according to it?
- template: Use `JkDocument`
- asset, copy
- jekyll site configuration, convert that

Reply:
Sorry, I didn't understand what you're meaning here :-/

Reply:
The change unconditionally runs files through this conversion process and erroring if they don't match up.

Some blogs will have assets, like image files.  We shouldn't attempt this process on them.

My guess is Jekyll uses a whitelist of file extensions with some being added via the [`markdown_ext`](https://jekyllrb.com/docs/configuration/#default-configuration) field.

In addition, some people will want their [configuration](https://jekyllrb.com/docs/configuration/) converted.

Reply:
Oh I see, this struct only manage post at the moment, maybe I should rename it accordingly and when we add more file type, create a struct for them too?

Reply:
I think the name works.  We could just add `JkConfig` for that.  For assets, we might just do a straight file copy.

At minimum, I'd appreciate it if the code would only apply `JkDocument` to the right type of files and `warn!` when it comes to unsupported files.

Reply:
Yup I understand better what you said, I'm processing only if extension is `md` or `markdown` for now. A JkConfig is a good idea as a second step, totally agree.